prov/tcp/src/tcpx_cq.c

```diff
@@ -35,19 +35,66 @@
 
 #include "tcpx.h"
 
+#define TCPX_DEF_CQ_SIZE (1024)
+
 static int tcpx_cq_close(struct fid *fid)
 {
 	int ret;
-	struct util_cq *cq;
+	struct tcpx_cq *tcpx_cq;
 
-	cq = container_of(fid, struct util_cq, cq_fid.fid);
-	ret = ofi_cq_cleanup(cq);
+	tcpx_cq = container_of(fid, struct tcpx_cq, util_cq.cq_fid.fid);
+	util_buf_pool_destroy(tcpx_cq->pe_entry_pool);
+	ret = ofi_cq_cleanup(&tcpx_cq->util_cq);
 	if (ret)
 		return ret;
-	free(cq);
+
+	free(tcpx_cq);
 	return 0;
 }
 
+struct tcpx_pe_entry *tcpx_pe_entry_alloc(struct tcpx_cq *tcpx_cq)
+{
+	struct tcpx_pe_entry *pe_entry;
+
+	fastlock_acquire(&tcpx_cq->util_cq.cq_lock);
+	pe_entry = util_buf_alloc(tcpx_cq->pe_entry_pool);
+	if (!pe_entry) {
+		fastlock_release(&tcpx_cq->util_cq.cq_lock);
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,"failed to get buffer\n");
+		return NULL;
+	}
+	memset(pe_entry, 0, sizeof(*pe_entry));
```

Comment:
This is cleared both on acquire and in the release call.

Reply:
removed in release